man/meson.build

```diff
@@ -235,6 +235,6 @@ update_man_rules = custom_target(
         output : 'update-man-rules',
         command : [sh, '-c',
                    'cd @0@ && '.format(project_build_root) +
-                   'python3 @0@/tools/update-man-rules.py $(find @0@ -wholename "*/man/*.xml") >t && '.format(project_source_root) +
+                   python.path() + ' @0@/tools/update-man-rules.py $(find @0@ -wholename "*/man/*.xml") >t && '.format(project_source_root) +
                    'mv t @0@/rules/meson.build'.format(meson.current_source_dir())],
```

Comment:
This is incorrect. It embeds the resolved python command into a complicated shell script, but does so incorrectly because that interpolation needs to be shell-quoted.

Admittedly, project_build_root and project_source_root are not shell-quoted and are both more likely to be broken than the path to python. After all, the path to python is generally in /usr/bin while the path to the source or build tree may trivially have whitespace in it.

Traditionally, one would pass all 3 strings as argv to the sh command (remembering to fill in a dummy arg for argv0), then use `"$1"` etc.

This highlights the fact that really, the entire command here looks quite strange. Why can't tools/update-man-rules.py be rewritten to find and iterate over xml files itself, and write the `meson.build` file directly once complete?

Reply:
Two advantages: this version will log the full list of files in the build logs, and it is also fairly easy to call update-man-rules with some subset of files for debugging/testing. We *could* rewrite it to make update-man-rules.py look for files itself, but as @poettering said, there is still no rationale what this patch is trying to fix.

Reply:
Well, there is the broken terribad shell quoting.

If you just wanted to log the files, you could have it accept the filesnames as argv when given an interactive `--files` argument or something, and otherwise do globbing inside python while using `print()` to perform logging.

I'm a bit confused about this logging thing actually, since it isn't being logged in the command line contained in build.ninja, so what is supposed to log the filenames? sh itself in xtrace mode? The script using `print()`?

One bonus if you simplify it to a single python script which does work internally, is you can just find_program(scriptname) and run that as is. Meson will find the correct python interpreter for you, and if all else fails it will use its own sys.executable.

Reply:
OK. I'll change the program to glob internally then.

Reply:
https://github.com/systemd/systemd/pull/22829